File: Data Structures & Algorithms/reverse-a-linked-list/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        //approach 1
        // if(head==null || head.next==null){
        //     return head;
        // }
        // ListNode temp=head;
        // Stack<Integer> stack =new Stack(); 
        // while(temp!=null){
        //     stack.push(temp.val);
        //     temp=temp.next;
        // }
        // temp=head;
        // while(temp!=null){
        //     int num = stack.pop();
        //     temp.val=num;
        //     temp=temp.next;
        // }
        // return head;

        // approach 2
        if(head==null || head.next==null){
            return head;
        }
        ListNode temp= head;
        ListNode prev=null;
        while(temp!=null){
            ListNode next = temp.next;
            temp.next=prev;
            prev=temp;
            temp=next;
        }
        return prev;
    }
}

File: Data Structures & Algorithms/remove-duplicates-from-sorted-array/submission-2.py
class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        # # approach 1
        # d={}
        # for num in nums:
        #     d[num]=0
        
        # i=0
        # for key in d:
        #     nums[i]=key
        #     i+=1
        # return i

        # approach 2

        if len(nums)<=1:
            return len(nums)
        

        l,r=0,1

        while r<len(nums):
            if nums[l]==nums[r]:
                r+=1
            elif r-1 != l:
                nums[l+1],nums[r]=nums[r],nums[l+1]
                r+=1
                l+=1
            else:
                l+=1
                r+=1
        return l+1
        
            




    # 1,1,1,1,1,2,2,3,3
    # l         r

    # 1,2,3,4,5
    #   l,r   
            
